### PR TITLE
Chore: Pass signed `user_hash` to Intercom via Rudderstack

### LINKED ISSUE #63749: Pass signed Intercom identityVerification to Rudderstack
In HG Intercom is loaded via Rudderstack, and Intercom requires a signed user hash to be specified for identity verification.

This 'user_hash' must be the ID we identify the user to rudderstack with, signed on the backend via a configured secret.

See grafana-com PR for implementation https://github.com/grafana/grafana-com/pull/5797, and context: https://raintank-corp.slack.com/archives/C13QEUGLB/p1677092852046129 via https://raintank-corp.slack.com/archives/C040HLM1Z7V/p1673050557093749

Note:

At the moment, the frontend decides the user identifier based on some logic https://github.com/grafana/grafana/blob/main/public/app/core/services/echo/utils.ts#L9-L15 but the identifier must match what the backend uses for the user_hash, so this logic should probably be moved to the backend, and the backend passes through (in CurrentUserDTO? config? A new object?) this both this opaque 'analytics identifier' and signed user_hash to the frontend.

This should be done in a reasonably analytics-service-agnostic way

### PR DESCRIPTION
**Pull request by ashharrison90**
**What is this feature?**

- moves analytics identifiers to the backend
- exposes them to the frontend via the `User` object

**Why do we need this feature?**

- need to pass a hashed user_id to intercom to verify they're the user we think they are

**Who is this feature for?**

- grafana devs

**Which issue(s) does this PR fix?**:

Fixes #63749 

**Special notes for your reviewer**:

context here: https://raintank-corp.slack.com/archives/C13QEUGLB/p1677092852046129

### CODE REVIEW

File: pkg/services/user/userimpl/user.go [hunk 1]
```diff
@@ -594,3 +598,25 @@ func (s *Service) supportBundleCollector() supportbundles.Collector {
 		Fn:                collectorFn,
 	}
 }
+
+func hashUserIdentifier(identifier string, secret string) string {
+	key := []byte(secret)
+	h := hmac.New(sha256.New, key)
+	h.Write([]byte(identifier))
+	return hex.EncodeToString(h.Sum(nil))
+}
```
Comment (ashharrison90) [inline pkg/services/user/userimpl/user.go:607]: taken from https://gist.github.com/thewheat/7342c76ade46e7322c3e#go (linked from the [intercom docs](https://www.intercom.com/help/en/articles/183-enable-identity-verification-for-web-and-mobile))

File: pkg/services/user/userimpl/user.go [hunk 2]
```diff
@@ -594,3 +598,25 @@ func (s *Service) supportBundleCollector() supportbundles.Collector {
...
+func buildUserAnalyticsSettings(signedInUser user.SignedInUser) user.AnalyticsSettings {
+	var settings user.AnalyticsSettings
+
+	if signedInUser.ExternalAuthID != "" {
+		settings.Identifier = signedInUser.ExternalAuthID
+	} else {
+		settings.Identifier = signedInUser.Email
+	}
+
+	if os.Getenv("GF_ANALYTICS_SECRET") != "" {
```
Comment (ashharrison90) [inline pkg/services/user/userimpl/user.go:618]: 2 questions:
- is this the best way to get a secret from the environment? 
- what should this be called? is `GF_ANALYTICS_SECRET` ok?

Comment (sakjur) [reply pkg/services/user/userimpl/user.go:618]: No, we don't read the environment directly but rather read from `pkg/setting/setting.go`

In your case, you should add something like `intercom_secret` to

https://github.com/grafana/grafana/blob/27635e6f7bbbe04cd9fdfdca9feacf97267145e5/pkg/setting/setting.go#L395-L403

and

https://github.com/grafana/grafana/blob/27635e6f7bbbe04cd9fdfdca9feacf97267145e5/pkg/setting/setting.go#L1029-L1036

And to the `defaults.ini` and then copy over to `sample.ini` and prefix the non-comment lines with `;` (if that sounds a bit unclear, compare the two files visually and it should be clear) somewhere here

https://github.com/grafana/grafana/blob/27635e6f7bbbe04cd9fdfdca9feacf97267145e5/conf/defaults.ini#L220

`GF_SECTION_CONFIGURATION_OPTION` will automatically be created from reading the `defaults.ini`.

File: pkg/services/user/userimpl/user.go [hunk 3]
```diff
@@ -594,3 +598,25 @@ func (s *Service) supportBundleCollector() supportbundles.Collector {
...
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func buildUserAnalyticsSettings(signedInUser user.SignedInUser) user.AnalyticsSettings {
+	var settings user.AnalyticsSettings
+
+	if signedInUser.ExternalAuthID != "" {
+		settings.Identifier = signedInUser.ExternalAuthID
+	} else {
+		settings.Identifier = signedInUser.Email
```
[the remark this answers is not in the document]
Comment (sakjur) [reply pkg/services/user/userimpl/user.go:615]: Stacks isn't a Grafana concept, though I guess the combination of `InstanceName` or `AppURL` from the `setting` package and the user's login/email could be considered to be semi-static and should be unique within an organization.

[the remark this answers is not in the document]
Comment (ashharrison90) [reply pkg/services/user/userimpl/user.go:615]: @sakjur believe it needs to be unique globally, not just within an organization. will that still hold true with `AppURL` + email? 🤔

[the remark this answers is not in the document]
Comment (sakjur) [reply pkg/services/user/userimpl/user.go:615]: @ashharrison90 No, but different organizations (to be clear, I'm not talking about orgs as defined in Grafana or HG but rather the hosting company, Grafana Labs for all of Hosted Grafana for example) should have different Intercom accounts, so possible overlap shouldn't be a problem? …and even then it should be unlikely for well-configured Grafana installations.

There'll be a lot of `http://localhost:3000` AppURLs/InstanceNames out there, but if you're interested in analytics, I think it's fair to say that that depends on you having different URLs for different installations of Grafana.

[the remark this answers is not in the document]
Comment (ashharrison90) [reply pkg/services/user/userimpl/user.go:614]: @sakjur yep makes sense. for hosted grafana, afaik we've got 1 global intercom account so as long as we've got something that's unique for each user to hosted grafana, that's good. 

the reason it's a bit more important that it's definitely unique is that i believe intercom is used for more than just analytics. think support chat bots etc. so we can't have anyone being able to impersonate another user in any way.

File: pkg/services/user/userimpl/user.go [hunk 4]
```diff
@@ -594,3 +597,25 @@ func (s *Service) supportBundleCollector() supportbundles.Collector {
...
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func buildUserAnalyticsSettings(signedInUser user.SignedInUser, intercomSecret string) user.AnalyticsSettings {
+	var settings user.AnalyticsSettings
+
+	if signedInUser.ExternalAuthID != "" {
+		settings.Identifier = signedInUser.ExternalAuthID
+	} else {
+		settings.Identifier = signedInUser.Email + "@" + setting.AppUrl
```
Comment (joshward) [inline pkg/services/user/userimpl/user.go:614]: re-thinking about this, I'm not sure we gain any value by adding the `AppUrl` to this. It's probably just safe to use Email as Identifier when we don't have the ExternalAuthID. Which could simplify this code back down.

unless you can think of a security reason to try to treat the user with the same email as different.

File: packages/grafana-data/src/types/config.ts [hunk 5]
```diff
@@ -96,6 +96,11 @@ export type OAuth =
  */
 export type OAuthSettings = Partial<Record<OAuth, { name: string; icon?: IconName }>>;
 
+export interface AnalyticsSettings {
```
Comment (tskarhed) [inline packages/grafana-data/src/types/config.ts:99]: Should we add a docstring to this?

**Review by sakjur — APPROVED** [review]
(BE only, changes looks good)

**Review by JoaoSilvaGrafana — APPROVED** [review]
Lgtm!

**Pull request merged**